tools/clang/lib/Sema/SemaHLSL.cpp

```diff
@@ -11534,34 +11607,35 @@ void hlsl::DiagnoseTranslationUnit(clang::Sema *self) {
       }
       pPatchFnDecl = NL.Found;
     }
-  }
-
-  if (pPatchFnDecl) {
-    FunctionDecl *patchResult = ValidateNoRecursion(self, pPatchFnDecl);
 
-    // In this case, recursion was detected in the patch-constant function
-    if (patchResult) {
-      self->Diag(patchResult->getSourceRange().getBegin(),
-                 diag::err_hlsl_no_recursion)
-          << 2 << patchResult->getName();
-    }
+    if (pPatchFnDecl) {
+      FunctionDecl *patchResult = ValidateNoRecursion(self, pPatchFnDecl);
 
-    // The patch function decl and the entry function decl should be
-    // disconnected with respect to the call graph.
-    // Only check this if neither function decl is recursive
-    if (!result && !patchResult) {
-      hlsl::CallGraphWithRecurseGuard CG;
-      CG.BuildForEntry(pPatchFnDecl);
-      if (CG.CheckReachability(pPatchFnDecl, pEntryPointDecl)) {
-        self->Diag(pEntryPointDecl->getSourceRange().getBegin(),
-                   diag::err_hlsl_patch_reachability_not_allowed)
-            << 1 << pEntryPointDecl->getName() << 0 << pPatchFnDecl->getName();
+      // In this case, recursion was detected in the patch-constant function
+      if (patchResult) {
+        self->Diag(patchResult->getSourceRange().getBegin(),
```

Comment:
Earlier, we use `DiagnosedDecls` to prevent duplicate recursion diagnostics for the same function.  We should do the same here.  That will also solve the issue that multiple hull shaders could use the same patch constant function, leading to duplicate recursion diagnostics for that patch constant function.

Reply:
Good point!